### PR TITLE
Error/Information/Warning using strings as id's

### PR DESCRIPTION
**Pull request by janezd**
When implementing a widget which issues multiple error, warning or information messages, we have to keep track of the used id's or pick random numbers and hope for the best to avoid collisions.

This PR proposes the use of Msg instances as id's.

The `OWWidget` class has local classes `Error`, `Warning`, `Information` that contains message instances that can be triggered or cleared.

The old methods `error`, `warning` and `information` can trigger a specific (protected) message instance. Using the old methods with id's is deprecated and triggers a warning.

Most widgets will require no other changes than removing the first argument in the call of `error`, `warning` or `information`. If they didn't use the id argument, no change is required.

### CODE REVIEW

File: Orange/widgets/evaluate/owtestlearners.py [hunk 1]
```diff
@@ -258,28 +279,28 @@ def set_train_data(self, data):
...
-            self.error(0, "Train data input requires a class variable")
+            self.message(self.ERR_CLASS_REQUIRED)
             data = None
 
         if isinstance(data, SqlTable):
             if data.approx_len() < AUTO_DL_LIMIT:
                 data = Table(data)
             else:
-                self.information(0, "Train data has been sampled")
+                self.message(self.INFO_DATA_SAMPLED)
```
Comment (kernc) [inline Orange/widgets/evaluate/owtestlearners.py]: Why then not just `self.INFO_DATA_SAMPLED.show()` and `self.WARN_MISSING_DATA.hide()` ?

**Comment by janezd** [comment]
The code is still somewhat broken, but could you (in particular @astaric) please comment on https://github.com/janezd/orange3/blob/11850947d79d575f00416f503e4c0aa4a562b623/doc/development/source/tutorial-utilities.rst#issuing-warning-and-errors. And on the code as well, of course.

The deeper I dig, the wider the scope of changes. After I reimplemented messages as objects and grouped them into instances of `Error`, `Warning` and `Information`, there is no sense to have a separate dictionary with lists of active messages, to pass strings in signals for the canvas etc.

**Comment by janezd** [comment]
My master branch is a mess; rebasing doesn't help, either. The relevant changes are in the last three commits. Other commits that are present apparently do not change any files, so "Files changed" shows the correct changes, too.

File: doc/development/source/tutorial-utilities.rst [hunk 2]
```diff
@@ -92,38 +92,60 @@ displayed in the top row of the widget and also indicated in the schema.
...
-for a widget.
+    class Error(widget.OWWidget.Error):
+        no_continuous_features = Msg("No continuous features")
 
-Multiple error messages can be removed by calling the function with a
-list of id's, `self.warning([1, 13, 42])`.
 
-Since most widgets have only a single error message, the id can be omitted
-(in which case it defaults to 0). In this case, ::
+    class Warning(widget.OWWidget.Error):
```
Comment (astaric) [inline doc/development/source/tutorial-utilities.rst]: class Warning(OWWidget.Warning)?

File: doc/development/source/tutorial-utilities.rst [hunk 3]
```diff
@@ -92,38 +92,60 @@ displayed in the top row of the widget and also indicated in the schema.
...
+If widget can issue multiple messages and some advanced handling is required,
+it needs to declare them as local classes within the widget class, and derive
+them from the corresponding `OWWidget` classes. For instance, a widget class
+can contain the following classes::
 
-A widget can have multiple such messages at the same time. They are
-distinguished by different id's, like `42` above. The id numbers are
-arbitrary and are assigned by the programmer. They need to be unique
-for a widget.
+    class Error(widget.OWWidget.Error):
```
Comment (astaric) [inline doc/development/source/tutorial-utilities.rst]: I would import OWWidget directly and use that (shorter) syntax in the example.

File: Orange/widgets/widget.py [hunk 4]
```diff
@@ -127,13 +133,14 @@ class OWWidget(QDialog, Report, ProgressBarMixin, metaclass=WidgetMetaClass):
...
     resizing_enabled = True
 
-    widgetStateChanged = Signal(str, int, str)
     blockingStateChanged = Signal(bool)
     processingStateChanged = Signal(int)
 
-    # For reasons I don't understand, the signal has to be defined here and
-    # not in the mix-in class, otherwise PyQt can't connect to it.
+    # For reasons I don't understand, these signals must be defined here and
+    # not in the mix-in classes, otherwise PyQt can't connect to it.
```
Comment (astaric) [inline Orange/widgets/widget.py]: *them

File: Orange/widgets/visualize/owheatmap.py [hunk 5]
```diff
@@ -681,8 +688,9 @@ def set_dataset(self, data=None):
                 self.error(0, "No continuous feature columns")
                 input_data = data = None
             else:
-                self.information(1, "{} discrete column{} ignored"
-                                .format(ndisc, "s" if ndisc > 1 else ""))
+                print("disc ignored")
```
Comment (astaric) [inline Orange/widgets/visualize/owheatmap.py]: Should this print be here?

File: Orange/widgets/evaluate/owtestlearners.py [hunk 6]
```diff
@@ -166,6 +166,25 @@ class OWTestLearners(widget.OWWidget):
...
 
+    class Error(OWWidget.Error):
+        class_required = Msg("Train data input requires a class variable")
+        class_required_test = Msg("Test data input requires a class variable")
+        too_many_folds = Msg("Number of folds exceeds the data size")
+        class_inconsistent = Msg("Test and train data sets "
+                                 "have different classes")
+
+    class Warning(OWWidget.Warning):
+        missing_data = Msg()
```
Comment (astaric) [inline Orange/widgets/evaluate/owtestlearners.py]: I cannot decide whether I like this short syntax or not. It is shorter to write than explicitly setting "{}" as a message, but I am not sure it is obvious that the message will be supplied later.

File: doc/development/source/tutorial-utilities.rst [hunk 7]
```diff
@@ -92,38 +92,60 @@ displayed in the top row of the widget and also indicated in the schema.
...
-Messages, warning and errors are issued by methods
-`self.information(id, text)`, `self.warning(id, text)` and
-`self.error(id, text)`, respectively. A warning like the one above is
-issued by ::
+Simple messages
+...............
 
-    self.warning(42, "Data contains continuous variables. "
-                 "Discretize the data to use them.")
+If the widget only issues a single type of error, warning and/or information,
```
Comment (astaric) [inline doc/development/source/tutorial-utilities.rst]: This section in the doc seems a bit condensed. Should we put examples in separate rows as in the next sections?

File: Orange/widgets/visualize/owheatmap.py (hunk 5 above)
Comment (janezd) [inline Orange/widgets/visualize/owheatmap.py]: No, this is `print`.

File: Orange/widgets/evaluate/owtestlearners.py (hunk 6 above)
Comment (janezd) [inline Orange/widgets/evaluate/owtestlearners.py]: Me neither. `Msg({})` doesn't look great, either. I'll define another class whose name will indicate that the whole message will be formed when it's raised, that is,

    missing_data = CustomMsg()

**Comment by janezd** [comment]
@astaric, do I have a green light to convert the widgets *en masse* to the new messages? Widgets (in particular in add-ons) that use old messages will work, but with warnings.

**Comment by astaric** [comment]
No, I would rather merge this PR and do the conversion in a new one. As it is, this PR is a pain to rebase, so I would try not to add many more commits to it :)

**Comment by astaric** [comment]
```
git checkout master
git reset --hard origin/master
git push -f
```
This will fix your master

```
git checkout widget-error-msgs
git rebase -i master
# Remove all irrelevant commits, then fix any conflicts encountered
git push -f
```

This will update this PR to include only relevant commits.

File: Orange/widgets/data/owdatasampler.py [hunk 8]
```diff
@@ -45,10 +45,10 @@ class OWDataSampler(widget.OWWidget):
     number_of_folds = Setting(10)
     selectedFold = Setting(1)
 
-    class Error(widget.OWWidget.Error):
-        too_many_folds = Msg("Number of folds exceeds the data size")
+    class Error(OWWidget.Error):
+        too_many_folds = Msg("Number of folds exceeds data size")
         sample_larger_than_data = Msg("Sample must be smaller than data")
-        not_enough_to_stratify = Msg("Not enough data for stratified sampling")
+        not_enough_to_stratify = Msg("Data is too small to stratify")
```
Comment (janezd) [inline Orange/widgets/data/owdatasampler.py]: These messages are shortened to fit into widgets.

**Comment by janezd** [comment]
I did both - the API and core widgets -  in the same PR: I was improving the API when modifying widgets. Changes in widgets are mostly benign, so rebasing shouldn't be a problem.

I haven't upgraded any add-ons. They should work (with deprecation warnings). Upgrading should be trivial, though.

This PR is ready for reviewing.

**Comment by astaric** [comment]
Please rebase onto current master. After that, the tests will fail, but I have prepared a fix: https://github.com/astaric/orange3/commit/7c82c3fa9e63e472aa398df200b4988bfd20f0cd

File: Orange/widgets/utils/owlearnerwidget.py [hunk 9]
```diff
@@ -108,8 +108,11 @@ class OWBaseLearner(OWWidget, metaclass=OWBaseLearnerMeta):
     resizing_enabled = False
     auto_apply = Setting(True)
 
-    DATA_ERROR_ID = 1
-    OUTDATED_LEARNER_WARNING_ID = 2
+    class Error(OWWidget.Error):
+        data_error = Msg("{}")
+
+    class Warning(OWWidget.Error):
```
Comment (astaric) [inline Orange/widgets/utils/owlearnerwidget.py]: (OWWidget.Warning)

File: doc/development/source/tutorial-utilities.rst [hunk 10]
```diff
@@ -92,38 +92,83 @@ displayed in the top row of the widget and also indicated in the schema.
...
-A widget can have multiple such messages at the same time. They are
-distinguished by different id's, like `42` above. The id numbers are
-arbitrary and are assigned by the programmer. They need to be unique
-for a widget.
+    class Error(OWWidget.Error):
+        no_continuous_features = Msg("No continuous features")
 
-Multiple error messages can be removed by calling the function with a
-list of id's, `self.warning([1, 13, 42])`.
+    class Warning(OWWidget.Error):
```
Comment (astaric) [inline doc/development/source/tutorial-utilities.rst]: (OWWidget.Warning)

**Pull request merged**